compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -1112,8 +1125,31 @@ pub fn typeid_for_instance<'tcx>(
     mut instance: Instance<'tcx>,
     options: TypeIdOptions,
 ) -> String {
-    if matches!(instance.def, ty::InstanceDef::Virtual(..)) {
-        instance.args = strip_receiver_auto(tcx, instance.args)
+    if (matches!(instance.def, ty::InstanceDef::Virtual(..))
+        && instance.def_id() == tcx.lang_items().drop_in_place_fn().unwrap())
+        || matches!(instance.def, ty::InstanceDef::DropGlue(..))
+    {
+        // Adjust the type ids of DropGlues
+        //
+        // DropGlues may have indirect calls to one or more given types drop function. Rust allows
+        // for types to be erased to any trait object and retains the drop function for the original
+        // type, which means at the indirect call sites in DropGlues, when typeid_for_fnabi is
+        // called a second time, it only has information after type erasure and it could be a call
+        // on any arbitrary trait object. Normalize them to a synthesized Drop trait object, both on
+        // declaration/definition, and during code generation at call sites so they have the same
+        // type id and match.
+        //
+        // FIXME(rcvalle): This allows a drop call on any trait object to call the drop function of
+        //   any other type.
+        //
+        let self_ty = Ty::new_mut_ref(
+            tcx,
+            tcx.lifetimes.re_erased,
+            new_dynamic_trait(tcx, tcx.lang_items().drop_trait().unwrap(), List::empty()),
+        );
+        instance.args = tcx.mk_args_trait(self_ty, List::empty());
```

Comment:
This is setting the self type of `drop_in_place` to `&mut dyn Drop`, not `dyn Drop` like @maurer was suggesting above.

Reply:
Oops, fixed.

Reply:
You did not fix this. Now you're setting the self type to `&dyn Drop`. It needs to be `dyn Drop`, so that for [`fn drop_in_place<T>(_: *mut T)`](https://doc.rust-lang.org/nightly/std/ptr/fn.drop_in_place.html), `T = dyn Drop`.

Reply:
Oh, drop_in_place has a ref already.

Reply:
And [`Drop::drop`](https://doc.rust-lang.org/nightly/std/ops/trait.Drop.html#tymethod.drop)

Reply:
Wait, we're replacing self of the monomorphized versions of it here. If we replace &dyn Trait with dyn Trait, wouldn't it be incorrectly reflecting what the generic meant for the monomorphized versions of it?

Reply:
Huh? We're replacing a concrete `T` with `dyn Drop` to do type erasure here.

Reply:
Recall that `Self` is a type parameter, not the argument 0 of these functions. `Self` should be substituted with `dyn Drop` to erase the original self type (e.g. `Vec<i32>`). There's no reason to wrap that type in an additional ref, mutable or immutable.

Reply:
This block replaces self both at the declaration/definition (i.e., `matches!(instance.def, ty::InstanceDef::DropGlue(..)`) and the call sites (i.e., `matches!(instance.def, ty::InstanceDef::Virtual(..)) && Some(instance.def_id()) == tcx.lang_items().drop_in_place_fn()`). And similarly to what we're doing for regular trait objects (and going to do for closures), the declaration/definition need to match the call sites (i.e., by using `&dyn Trait`).

Reply:
Again, using `&dyn Trait` is not correct here.

Please inspect the contents of the `instance.args[0]` to convince yourself that the `Self` parameter of `Drop` and the `T` parameter of `drop_in_place` denote the *contents* of what is being dropped, and don't need to be populated with a reference. If we put `&dyn Drop` into `args[0]` here, we're creating an instance that corresponds dropping a *reference* to `&dyn Drop`, which is not correct.

Reply:
Reminder that `args` is not the call arguments of the function signature but *generic* arguments, which are used when substituting the generic parameters of the definition. Specifically, if we use `dyn Drop` here, we will be creating an instance that corresponds to the implementation `impl Drop for dyn Drop` and `std::mem::drop_in_place::<dyn Drop>`.

Reply:
You're right. We're replacing Instance args (i.e., pre substitution), not self in the FnAbi anymore, after I changed it to do that. That was my confusion.

Reply:
Done. I also added a codegen test for it (which I wanted to do already).